modules/social_features/social_core/src/Plugin/Field/FieldType/ExternalIdentifierItem.php

```diff
@@ -0,0 +1,221 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\social_core\Plugin\Field\FieldType;
+
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Field\FieldItemBase;
+use Drupal\Core\Field\FieldStorageDefinitionInterface;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\StringTranslation\TranslatableMarkup;
+use Drupal\Core\TypedData\DataDefinition;
+use Drupal\Core\TypedData\DataReferenceTargetDefinition;
+
+/**
+ * Defines the 'social_external_identifier' field type.
+ *
+ * Supported settings (below the definition's 'settings' key) are:
+ * - target_type: The entity type to reference. Required.
+ *
+ * @FieldType(
+ *   id = "social_external_identifier",
+ *   label = @Translation("External Identifier"),
+ *   description = @Translation("Define external Identifier and external Owner/Consumer."),
+ *   default_widget = "social_external_identifier_default_widget",
+ *   default_formatter = "social_external_identifier_default_formatter",
+ *   constraints = {
+ *     "ExternalIdentifierEmptySubfieldsConstraint" = {},
+ *     "ExternalIdentifierExternalOwnerTargetTypeConstraint" = {},
+ *     "ExternalIdentifierExternalOwnerIdConstraint" = {},
+ *     "ComplexData" = {
+ *       "external_id" = {
+ *         "Length" = {"max" = 225}
+ *        }
+ *      }
+ *   }
+ * )
+ */
+final class ExternalIdentifierItem extends FieldItemBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function isEmpty(): bool {
+    // If at least one of 3 subfields is not empty, field is not empty.
+    // Subfields are:
+    // - external_id,
+    // - external_owner_target_type
+    // - external_owner_id
```

Comment:
From a logic point of view I'm wondering if this is correct. Especially if we only expect the field to be valid when all 3 values are filled in, then shouldn't the field be considered empty if any of these values is empty (basically the inverse of the current logic)?

Reply:
Actually no :) 
- Field is considered empty if all values are empty
- Field is considered invalid not all values are are provided (except if all values are empty) - to reframe (if values are partly provided, field is considered as invalid (by constraints).
- Field is considered valid if all values are provided (but it is also valid if non of the value is provided)

Let me know if this is better explained now.

Reply:
That makes sense :D I think that ties in to the other thread we're discussing and if we have clarity there then I believe we cover all cases. Thanks!

Reply:
Ok I will update my explanation in the code with bullet points I've put in my previous comment, but let's resolve the other thread first ;)